[build tooling] Make build-webkit script respect output verbosity defaults

zakariaridouh opened the pull request:
#### a8ff9a266fcef1bf89db74d1562b83b190ce747f
<pre>
[build tooling] Make build-webkit script respect output verbosity defaults
<a href="https://bugs.webkit.org/show_bug.cgi?id=300210">https://bugs.webkit.org/show_bug.cgi?id=300210</a>
<a href="https://rdar.apple.com/162009595">rdar://162009595</a>

Reviewed by NOBODY (OOPS!).

We should respect the user defaults set when doing a build, a piping the output through filter-build-webkit when appropriate.

`defaults write org.webkit.BuildConfiguration BuildTranscriptVerbosity {quiet, oneline, default}`

We already had this logic in makefile.shared, so port it to the `build-webkit` script.

* Tools/Scripts/build-webkit:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a8ff9a266fcef1bf89db74d1562b83b190ce747f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/145175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/17856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/9631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/153847 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/98811 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/7ece860b-8fd7-4ac3-ae60-f17e06d167ea) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/147050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/18339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/17748 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/111625 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/80016 "2 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/7a337bad-5574-403c-b29b-a7f101950289) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/148138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/13988 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/130398 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/92523 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/b4e1706d-1541-4555-bd52-8aab4adf4c6f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/13348 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/11110 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/1292 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/137166 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/122871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/7156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/156159 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/5984 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/17707 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/8244 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/119633 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/17753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/14773 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/119967 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/15738 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/128414 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/73371 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/17328 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/6672 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/176464 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/17065 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/45387 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/17273 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/17128 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -286,7 +305,7 @@ if (isWin() || isPlayStation() || (isJSCOnly() && isWindows())) {
     exit 0 if isGenerateProjectOnly();
 
-    my @local_options = @options;
+    @local_options = @options;
```
Doesn't this just override whatever format arguments we've set?

(can be undone if you accept my suggestion to move the logic)

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -190,6 +191,24 @@ if ($showHelp) {
+    if ($verbosity eq "quiet" || $verbosity eq "oneline") {
+        my $filterScript = File::Spec->catfile(sourceDir(), "Tools", "Scripts", "filter-build-webkit");
+        $outputFilter = "| '$filterScript'";
+        $outputFilter .= " -f oneline" if $verbosity eq "oneline";
+    } elsif ($verbosity eq "default") {
+        push @local_options, "-hideShellScriptEnvironment";
+    }
+}
```
Could you move this logic to the `elsif (isAppleCocoaWebKit() && !isCMakeBuild()) ` branch starting on L306? I think you can get rid of your new `@local_options` variable by doing so, because you can just push to the local `@options` at that point.

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -299,12 +318,12 @@ if (isWin() || isPlayStation() || (isJSCOnly() && isWindows())) {
     # Build, and abort if the build fails.
     my $scheme = $xcodeScheme // "Everything up to WebKit + Tools";
-    $result = buildXcodeScheme($scheme, $clean, @local_options, @ARGV);
+    my $result = buildXcodeScheme($scheme, $clean, $outputFilter, @local_options, @ARGV);
```
One problem with this approach is that there will be no filtering for the `build-imagediff` invocation that happens below, which makes a second xcodebuild invocation.

I would suggest that instead of touching webkitdirs code, you spawn filter-build-webkit from within this subroutine and redirect STDOUT to it. Like this:

```perl
    my $filterBuildWebKit, my $oldout;
    if (!$verbose) {
        open ($oldout, ">&STDOUT") or die "can't dup STDOUT: $!";
        my $filterScript = File::Spec->catfile(sourceDir(), "Tools", "Scripts", "filter-build-webkit");
        my $filterPID = open ($filterBuildWebKit, "|-", $filterScript, @outputFilter);
        die "fork failed: $!" unless defined $filterPID;
        open (STDOUT, ">&", $filterBuildWebKit) or die "can't redirect STDOUT: $!";
    }
```

then after calling buildXcodeScheme and  `build-imagediff`:

```perl
    if (!$verbose) {
        open (STDOUT, ">&", $oldout) or die "can't restore STDOUT: $!";
        close $filterBuildWebKit or die "wait failed: $!";
    }
```

This also avoids having to using `bash -o pipefail`, which is kind of icky to me compared to just doing it natively in perl.

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -190,6 +191,24 @@ if ($showHelp) {
+    if ($verbosity eq "quiet" || $verbosity eq "oneline") {
+        my $filterScript = File::Spec->catfile(sourceDir(), "Tools", "Scripts", "filter-build-webkit");
+        $outputFilter = "| '$filterScript'";
+        $outputFilter .= " -f oneline" if $verbosity eq "oneline";
```
With my suggestion to use native perl IPC code, it makes more sense to store the output arguments as an array, like this:

```perl
my $verbosity = "default";
my @outputFilter;

...

    if ($verbosity eq "quiet" || $verbosity eq "oneline") {
        push @outputFilter, "-f", "oneline" if $verbosity eq "oneline";
    } elsif ...
```

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -190,6 +191,24 @@ if ($showHelp) {
    exit 1;
 }
 
+my $verbosity = "default";
```
Could we extend this to work with the existing `--verbose` flag? I'm imagining something like this in usage text:

```
-v, --verbose  Show verbose build output 
-q, --quiet    Filter Xcode build output to one line per task
--verbosity    Filter Xcode build output according to the given filter-build-webkit format ("default", "quiet", "oneline")

On macOS, the default verbosity can be persisted using `defaults write org.webkit.BuildConfiguration BuildTranscriptVerbosity <verbosity>`.
```

And in code:
```perl
my $verbosity = "default";
my $quiet = 0;

my %options = (
    ...
    'v|verbose' => \$verbose,
    'q|quiet' => \$quiet,
    'verbosity' => \$verbosity,
    ...
);

GetOptions(%options);

die "--verbose and --quiet are mutually exclusive.\n" if $verbose && $quiet;
$verbosity = "quiet" if $quiet;
```

emw-apple reviewed (changes requested):
Makes sense, and I appreciate working towards unifying makefile and build-webkit logic.

I hacked on the following suggestions myself, so if you'd like, I can push my changes to your branch.

zakariaridouh committed:
[build tooling] Make build-webkit script respect output verbosity defaults
https://bugs.webkit.org/show_bug.cgi?id=300210
rdar://162009595

Reviewed by NOBODY (OOPS!).

We should respect the user defaults set when doing a build, a piping the output through filter-build-webkit when appropriate.

`defaults write org.webkit.BuildConfiguration BuildTranscriptVerbosity {quiet, oneline, default}`

We already had this logic in makefile.shared, so port it to the `build-webkit` script.

* Tools/Scripts/build-webkit:
